crates/ide_assists/src/handlers/extract_function.rs

```diff
@@ -1441,21 +1448,31 @@ fn make_body(
         FunctionBody::Span { parent, text_range } => {
             let mut elements: Vec<_> = parent
                 .syntax()
-                .children()
+                .children_with_tokens()
                 .filter(|it| text_range.contains_range(it.text_range()))
-                .map(|it| rewrite_body_segment(ctx, &fun.params, &handler, &it))
+                .map(|it| match &it {
+                    syntax::NodeOrToken::Node(n) => syntax::NodeOrToken::Node(
+                        rewrite_body_segment(ctx, &fun.params, &handler, &n),
+                    ),
+                    _ => it,
+                })
                 .collect();
 
-            let mut tail_expr = match elements.pop() {
-                Some(node) => ast::Expr::cast(node.clone()).or_else(|| {
-                    elements.push(node);
-                    None
-                }),
-                None => None,
+            let mut tail_expr = match &elements.last() {
+                Some(element) => match element {
+                    syntax::NodeOrToken::Node(node) if ast::Expr::can_cast(node.kind()) => {
+                        ast::Expr::cast(node.clone())
+                    }
+                    _ => None,
+                },
+                _ => None,
             };
```

Comment:
```suggestion
            let mut tail_expr = match &elements.last() {
                Some(syntax::NodeOrToken::Node(node)) if ast::Expr::can_cast(node.kind()) => ast::Expr::cast(node.clone()),
                _ => None,
            };
```
Can be collapsed into one match